db/version_set.cc

```diff
@@ -5112,33 +5115,51 @@ Status VersionSet::ProcessManifestWrites(
 #endif  // NDEBUG
 
   // Assign and verify ManifestUpdateSequences
+  std::optional<std::string> pending_persist_replication_sequence;
+  uint64_t pending_manifest_update_sequence = manifest_update_sequence_;
+  bool is_leader{false};
+  ReplicationEpochAdditions new_replication_epochs;
   {
     Status s;
 
     for (auto& e : batch_edits) {
       if (!db_options_->replication_log_listener &&
-          !e->HasManifestUpdateSequence() && manifest_update_sequence_ == 0) {
+          !e->HasManifestUpdateSequence() && pending_manifest_update_sequence == 0) {
         // No physical replication, skip
         continue;
       }
 
-      ++manifest_update_sequence_;
+      ++pending_manifest_update_sequence;
 
       if (e->HasManifestUpdateSequence()) {
         // Update already has the manifest update sequence, verify that it's
         // correct
-        if (e->GetManifestUpdateSequence() != manifest_update_sequence_) {
+        if (e->GetManifestUpdateSequence() != pending_manifest_update_sequence) {
           std::ostringstream oss;
           oss << "Gap in ManifestUpdateSequence while writing to manifest, "
                  "expected="
-              << manifest_update_sequence_
+              << pending_manifest_update_sequence
               << " got=" << e->GetManifestUpdateSequence();
           s = Status::Corruption(oss.str());
           break;
         }
+
+        // replicating `ReplicationEpoch` additions from leader
+        for (const auto& epoch_addition : e->GetReplicationEpochAdditions()) {
+          new_replication_epochs.push_back(epoch_addition);
+        }
+        if (!s.ok()) {
+          break;
+        }
       } else if (db_options_->replication_log_listener) {
         // No manifest update sequence, this is the leader, set it.
-        e->SetManifestUpdateSequence(manifest_update_sequence_);
+        e->SetManifestUpdateSequence(pending_manifest_update_sequence);
+
+        is_leader = true;
```

Comment:
Keep in mind that the comment on L5155 is obsolete. When I first wrote the code only the leader had `replication_log_listener`, but that has since changed. I think we can completely kill `bool is_leader` -- only the leader will have `next_replication_epoch_` set.

Reply:
> only the leader will have next_replication_epoch_ set
that's not true. When follower applies `kNewEpoch` record, it will call `db_->NewManifestOnNextUpdate` as well. 

Also, L5155 might not be obsolete. It's leader because only leader doesn't have `manifest update sequence ` set.

Reply:
> that's not true. When follower applies kNewEpoch record, it will call db_->NewManifestOnNextUpdate as well.

Ah, okay, you're right.